src/compiler/checker.ts

```diff
@@ -12647,6 +12642,11 @@ namespace ts {
                     /*readonly*/ false, target.labeledElementDeclarations && target.labeledElementDeclarations.slice(index, endIndex));
         }
 
+        function getKnownKeysOfTupleType(type: TupleTypeReference) {
+            return getUnionType(append(arrayOf(type.target.fixedLength, i => getLiteralType("" + i)),

```

Comment:
If the tuple has a rest, shouldn't the keys contain `number`?

Reply:
The keys already contain `number` because all array types have an index signature.

Reply:
Hm. I question if we should be OK with `2` being assignable to `keyof [...T]` where `T extends [] | [object]`. But... I guess it's inline with how we already handle the keys of tuples?